R/shiny-options.R

```diff
@@ -8,31 +8,85 @@ getShinyOption <- function(name, default = NULL) {
+#' @description
+#'
+#' There are two mechanisms for working with options for Shiny. One is the
+#' [options()] function, which is part of base R, and the other is the
+#' `shinyOptions()` function, which is in the Shiny package. The reason for
+#' these two mechanisms is has to do with legacy code and scoping.
+#'
+#' The [options()] function sets options globally, for the duration of an R
+#' session. The [getOption()] function retrieves the value of an option.
+#'
+#' The `shinyOptions()` function sets the value of a shiny option, but unlike
+#' `options()`, it is not always global in scope; the options may be scoped
+#' globally, to an application, or to a user session in an application,
+#' depending on the context. The `getShinyOption()` function retrieves a value
+#' of a shiny option.
 #'
 #' @section Scope:
-#' There is a global option set which is available by default. When a Shiny
-#' application is run with [runApp()], that option set is duplicated and the
-#' new option set is available for getting or setting values. If options
-#' are set from `global.R`, `app.R`, `ui.R`, or `server.R`, or if they are set
-#' from inside the server function, then the options will be scoped to the
-#' application. When the application exits, the new option set is discarded and
-#' the global option set is restored.
-#'
-#' @section Options:
-#' There are a number of global options that affect Shiny's behavior. These can
-#' be set globally with `options()` or locally (for a single app) with
-#' `shinyOptions()`.
+#'
+#'   There are three levels of scoping for `shinyOptions()`: global,
+#'   application, and session.
+#'
+#'   The global option set is available by default. Any calls to
+#'   `shinyOptions()` and `getShinyOption()` outside of an app will access the
+#'   global option set.
+#'
+#'   When a Shiny application is run with [runApp()], the global option set is
+#'   duplicated and the new option set is available at the application level. If
+#'   options are set from `global.R`, `app.R`, `ui.R`, or `server.R` (but
+#'   outside of the server function), then the application-level options will be
+#'   modified.
+#'
+#'   Each time a user session is started, the application-level option set is
+#'   duplicated, for that session. If the options are set from inside the server
+#'   function, then they will be scoped to the session.
+#'
+#' @section Options with `shinyOptions()`:
+#'
+#'   There are a number of global options that affect Shiny's behavior. These
+#'   can be set globally with `options()` or locally (for a single app) with
+#'   `shinyOptions()`.
+#'
+#' \describe{
+#' \item{bootstraplib (defaults to `FALSE`)}{When a global **bootstraplib** theme
+#'   is set (via [bootstraplib::bs_theme_new()]), should **shiny** compile it
+#'   (via [bootstraplib::bootstrap()]) and include the result with [bootstrapPage()]
+#'   as well as use it to set better default styles for input widgets (e.g., [selectInput()])?}
+#' }
```

Comment:
This is a fantastic overhaul of the docs, thank you! Only nit here is that we likely won't be documenting a `bootstraplib` `shinyOption()`, so this section might end up being empty after https://github.com/rstudio/shiny/pull/3062, but maybe this provides motivation move the existing `getOption()` calls to `getShinyOption()` (as we made a note to do in https://github.com/rstudio/shiny/issues/2953)?

Reply:
I'll take out the `bootstraplib` docs for now, and I'll add in documentation for `cache`, since it should be there.

As for the `shinyOptions` and `options` fall-through, I can do that in another PR.

Reply:
Ohh, actually, `shiny.testmode` should be transferred to this section in this PR, so at least that gets rid of the empty section issue

Reply:
`shiny.testmode` actually should still be in the `options` section.

There's an internally-used shinyOption named `testmode`, but its purpose is actually to transfer the value of the `test.mode` argument from `runApp()` (as in `runApp(test.mode=T)`) to both the UI and server code. That was one of the early uses of `shinyOptions`, where the purpose was to capture some information at the application level and share it with both the UI and server code.

I could change the name of the internally-used shinyOption `testmode` to disambiguate it from the global option `shiny.testmode`.

Reply:
Although, now that I look at it, I think we can actually remove the `test.mode` option from `runApp()`, since shinytest doesn't actually use it. I'll merge this and then clean all that up in another PR.